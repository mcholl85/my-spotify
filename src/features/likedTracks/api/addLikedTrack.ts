import { useMutation, useQueryClient } from '@tanstack/react-query'
import api from '../../../lib/axios'
import { QUERY_LIKED_TRACKS } from '../../../constants/query.constants'

const addLikedTrack = async (id: string) => {
  await api.put(`/me/tracks?ids=${id}`)
}

export const useAddLikedTrack = () => {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: addLikedTrack,
    onSuccess: () => {
      queryClient.invalidateQueries([QUERY_LIKED_TRACKS])
    },
  })
}
