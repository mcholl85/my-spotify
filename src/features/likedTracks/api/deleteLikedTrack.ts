import { useMutation, useQueryClient } from '@tanstack/react-query'
import api from '../../../lib/axios'
import { QUERY_LIKED_TRACKS } from '../../../constants/query.constants'
import { QueryData } from '../types/queryData'
import { changeLikeStatus } from '../utils/changeLikeStatus'

const deleteLikedTrack = async (id: string) => {
  await api.delete(`/me/tracks?ids=${id}`)
}

type useDeleteLikedTrackProps = {
  offset: number
}

export const useDeleteLikedTrack = ({ offset }: useDeleteLikedTrackProps) => {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: deleteLikedTrack,
    onSuccess: (_, id) =>
      queryClient.setQueryData([QUERY_LIKED_TRACKS, offset], (oldData: QueryData) =>
        changeLikeStatus({ data: oldData, id, isLike: false }),
      ),
  })
}
