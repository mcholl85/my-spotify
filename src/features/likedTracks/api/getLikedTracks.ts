import api from '../../../lib/axios'
import { Track } from '../types/track'
import { useQuery } from '@tanstack/react-query'
import { QUERY_LIKED_TRACKS } from '../../../constants/query.constants'
import { DEFAULT_LIMIT, DEFAULT_OFFSET } from '../../../constants/api.constants'
import { setTrackList } from '../utils/setTrackList'

type GetLikedTracksParams = {
  limit?: number
  offset?: number
}

type useLikedTracksProps = GetLikedTracksParams

export type GetLikedTracksResponse = {
  href: string
  limit: number
  next: string
  offset: number
  previous: string
  total: number
  items: {
    added_at: Date
    track: Track
  }[]
}

const getLikedTracks = async ({
  limit = DEFAULT_LIMIT,
  offset = DEFAULT_OFFSET,
}: GetLikedTracksParams): Promise<GetLikedTracksResponse> => {
  return await (
    await api.get(`/me/tracks?limit=${limit}&offset=${offset}`)
  ).data
}

export const useLikedTracks = ({ limit, offset }: useLikedTracksProps) => {
  return useQuery({
    queryKey: [QUERY_LIKED_TRACKS, offset],
    queryFn: async () => {
      const data = await getLikedTracks({ limit, offset })
      return { total: data.total, tracks: setTrackList(data) }
    },
    retry: 0,
    refetchOnWindowFocus: false,
  })
}
