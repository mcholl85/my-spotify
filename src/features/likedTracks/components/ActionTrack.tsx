import { useEffect, useState } from 'react'
import { useAddLikedTrack } from '../api/addLikedTrack'
import { useDeleteLikedTrack } from '../api/deleteLikedTrack'
import { Button } from 'antd'

type ActionTrackProps = {
  id: string
}

export const ActionTrack = ({ id }: ActionTrackProps) => {
  const {
    mutate: addLikedTrack,
    isSuccess: addLikedTrackIsSuccess,
    isLoading: addLikedTrackIsLoading,
  } = useAddLikedTrack()
  const {
    mutate: removeLikedTrack,
    isSuccess: removeLikedTrackIsSuccess,
    isLoading: removeLikedTrackIsLoading,
  } = useDeleteLikedTrack()
  const [trackIsDelete, setTrackIsDelete] = useState(false)

  useEffect(() => {
    if (addLikedTrackIsSuccess) setTrackIsDelete(false)
  }, [addLikedTrackIsSuccess])

  useEffect(() => {
    if (removeLikedTrackIsSuccess) setTrackIsDelete(true)
  }, [removeLikedTrackIsLoading])

  return !trackIsDelete ? (
    <Button type='default' loading={removeLikedTrackIsLoading} onClick={() => removeLikedTrack(id)}>
      Enlever de Titres likés
    </Button>
  ) : (
    <Button type='primary' loading={addLikedTrackIsLoading} onClick={() => addLikedTrack(id)}>
      Ajouter aux Titres likés
    </Button>
  )
}
