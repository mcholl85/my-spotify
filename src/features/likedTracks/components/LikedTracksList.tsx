import Table, { ColumnsType } from 'antd/es/table'
import { Result, Space, Spin } from 'antd'
import { useLikedTracks } from '../api/getLikedTracks'
import { ActionTrack } from './ActionTrack'
import { setTrackList } from '../utils/setTrackList'
import { Column } from '../types/colum'

export const LikedTracksList = () => {
  const { data, isLoading, isError } = useLikedTracks({})
  const nbOfTracks = data?.items.length
  const columns: ColumnsType<Column> = [
    {
      title: 'Album',
      dataIndex: 'album',
      key: 'album',
      render: (_, record) => <img src={record.cover} alt={`${record.name} cover`} />,
    },
    {
      title: 'Titre',
      dataIndex: 'name',
      key: 'name',
    },
    {
      title: 'Action',
      key: 'action',
      render: (_, record) => <ActionTrack id={record.id} />,
    },
  ]

  if (isLoading)
    return (
      <Space size='large' className='container'>
        <Spin size='large' />
        <Space>Chargement en cours</Space>
      </Space>
    )
  if (isError) return <Result status='403' title='403' subTitle="Le token d'accès a expirée" />

  return (
    <Space direction='vertical' className='container'>
      <Result status='success' title={`Les ${nbOfTracks} musiques Spotify ont été chargées`} />
      <Table columns={columns} dataSource={setTrackList(data)} pagination={false} />
    </Space>
  )
}
