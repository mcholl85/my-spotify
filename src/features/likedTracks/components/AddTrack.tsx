import { useAddLikedTrack } from '../api/addLikedTrack'
import { DeleteTrack } from './DeleteTrack'

type AddTrackProps = {
  id: string
}

export const AddTrack = ({ id }: AddTrackProps) => {
  const { mutate: addLikedTrack, isSuccess } = useAddLikedTrack()

  if (isSuccess) return <DeleteTrack id={id} />

  return <button onClick={() => addLikedTrack(id)}>Ajouter aux Titres lik√©s</button>
}
