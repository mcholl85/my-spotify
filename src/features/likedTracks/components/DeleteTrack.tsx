import { useDeleteLikedTrack } from '../api/deleteLikedTrack'
import { AddTrack } from './AddTrack'

type DeleteTrackProps = {
  id: string
}

export const DeleteTrack = ({ id }: DeleteTrackProps) => {
  const { mutate: removeLikedTrack, isSuccess } = useDeleteLikedTrack()

  if (isSuccess) return <AddTrack id={id} />

  return <button onClick={() => removeLikedTrack(id)}>Enlever de Titres lik√©s</button>
}
