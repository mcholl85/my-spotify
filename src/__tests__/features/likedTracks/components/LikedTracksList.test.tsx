import { render, screen } from '@testing-library/react'
import { useLikedTracks } from '../../../../features/likedTracks/api/getLikedTracks'
import { LikedTracksList } from '../../../../features/likedTracks/components/LikedTracksList'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import userEvent from '@testing-library/user-event'
import { PropsWithChildren } from 'react'

jest.mock('../../../../features/likedTracks/api/getLikedTracks', () => ({
  useLikedTracks: jest.fn(),
}))

const queryClient = new QueryClient()
const wrapper = ({ children }: PropsWithChildren) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
)

const mockUseLikedTracks = useLikedTracks as jest.Mock

describe('Component LikedTracksList', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    })
  })
  afterEach(() => {
    jest.clearAllMocks()
  })
  test('should match DOM snapshots', () => {
    mockUseLikedTracks.mockReturnValue({
      data: {
        tracks: [{ key: 0, name: 'track1', cover: '', id: '123', liked: true }],
        total: 1,
      },
      isLoading: false,
      isError: false,
    })

    const { asFragment } = render(<LikedTracksList />, { wrapper })

    expect(asFragment()).toMatchSnapshot()
  })
  test('should render a 403 given isError is true', () => {
    mockUseLikedTracks.mockReturnValue({ data: {}, isLoading: false, isError: true })

    render(<LikedTracksList />)

    expect(screen.getByText('403')).toBeInTheDocument()
  })
  test('should render a loader given isLoading is true', () => {
    mockUseLikedTracks.mockReturnValue({ data: {}, isLoading: true, isError: false })

    render(<LikedTracksList />)

    expect(screen.getByText('Chargement en cours')).toBeInTheDocument()
  })
  test('should render a success message given isSuccess is true', () => {
    mockUseLikedTracks.mockReturnValue({
      data: { total: 2, tracks: [] },
      isLoading: false,
      isError: false,
    })

    render(<LikedTracksList />)

    expect(screen.getByText('Les 2 musiques Spotify ont été chargées')).toBeInTheDocument()
  })
  test.each`
    action      | text                          | liked
    ${'delete'} | ${'Enlever de Titres likés'}  | ${true}
    ${'add'}    | ${'Ajouter aux Titres likés'} | ${false}
  `('should render a $action button given the liked property is $liked ', ({ text, liked }) => {
    mockUseLikedTracks.mockReturnValue({
      data: {
        tracks: [{ key: 0, name: 'track1', cover: '', id: '123', liked }],
        total: 1,
      },
      isLoading: false,
      isError: false,
    })

    render(<LikedTracksList />, { wrapper })

    expect(screen.getByRole('button', { name: text })).toBeInTheDocument()
  })
  test('should render the track name and the cover image', () => {
    mockUseLikedTracks.mockReturnValue({
      data: {
        tracks: [{ key: 0, name: 'track1', cover: 'http://fakeUrl/', id: '123', liked: true }],
        total: 1,
      },
      isLoading: false,
      isError: false,
    })

    render(<LikedTracksList />, { wrapper })

    expect(screen.getByAltText('track1 cover')).toBeInTheDocument()
    expect(screen.getByText('track1')).toBeInTheDocument()
  })
  test('should load a new list of tracks given a click on the next page', async () => {
    mockUseLikedTracks.mockReturnValue({
      data: {
        tracks: [{ key: 0, name: 'track1', cover: 'http://fakeUrl/', id: '123', liked: true }],
        total: 51,
      },
      isLoading: false,
      isError: false,
    })

    render(<LikedTracksList />, { wrapper })
    expect(mockUseLikedTracks).toBeCalledWith({ offset: 0 })

    await userEvent.click(screen.getByText('2'))

    expect(mockUseLikedTracks).toBeCalledTimes(2)
    expect(mockUseLikedTracks).toBeCalledWith({ offset: 50 })
  })
})
