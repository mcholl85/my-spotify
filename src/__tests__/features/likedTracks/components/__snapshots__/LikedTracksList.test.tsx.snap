// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component LikedTracksList should match DOM snapshots 1`] = `
<DocumentFragment>
  <div
    class="ant-space css-dev-only-do-not-override-12jzuas ant-space-vertical container"
  >
    <div
      class="ant-space-item"
      style="margin-bottom: 8px;"
    >
      <div
        class="ant-result ant-result-success css-dev-only-do-not-override-12jzuas"
      >
        <div
          class="ant-result-icon"
        >
          <span
            aria-label="check-circle"
            class="anticon anticon-check-circle"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="check-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm193.5 301.7l-210.6 292a31.8 31.8 0 01-51.7 0L318.5 484.9c-3.8-5.3 0-12.7 6.5-12.7h46.9c10.2 0 19.9 4.9 25.9 13.3l71.2 98.8 157.2-218c6-8.3 15.6-13.3 25.9-13.3H699c6.5 0 10.3 7.4 6.5 12.7z"
              />
            </svg>
          </span>
        </div>
        <div
          class="ant-result-title"
        >
          Les 1 musiques Spotify ont été chargées
        </div>
      </div>
    </div>
    <div
      class="ant-space-item"
      style="margin-bottom: 8px;"
    >
      <ul
        class="ant-pagination css-dev-only-do-not-override-12jzuas"
      >
        <li
          aria-disabled="true"
          class="ant-pagination-prev ant-pagination-disabled"
          title="Previous Page"
        >
          <button
            class="ant-pagination-item-link"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              aria-label="left"
              class="anticon anticon-left"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="left"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                />
              </svg>
            </span>
          </button>
        </li>
        <li
          class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
          tabindex="0"
          title="1"
        >
          <a
            rel="nofollow"
          >
            1
          </a>
        </li>
        <li
          aria-disabled="true"
          class="ant-pagination-next ant-pagination-disabled"
          title="Next Page"
        >
          <button
            class="ant-pagination-item-link"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              aria-label="right"
              class="anticon anticon-right"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="right"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                />
              </svg>
            </span>
          </button>
        </li>
      </ul>
    </div>
    <div
      class="ant-space-item"
    >
      <div
        class="ant-table-wrapper css-dev-only-do-not-override-12jzuas"
      >
        <div
          class="ant-spin-nested-loading css-dev-only-do-not-override-12jzuas"
        >
          <div
            class="ant-spin-container"
          >
            <div
              class="ant-table"
            >
              <div
                class="ant-table-container"
              >
                <div
                  class="ant-table-content"
                >
                  <table
                    style="table-layout: auto;"
                  >
                    <colgroup />
                    <thead
                      class="ant-table-thead"
                    >
                      <tr>
                        <th
                          class="ant-table-cell"
                          scope="col"
                        >
                          Album
                        </th>
                        <th
                          class="ant-table-cell"
                          scope="col"
                        >
                          Titre
                        </th>
                        <th
                          class="ant-table-cell"
                          scope="col"
                        >
                          Action
                        </th>
                      </tr>
                    </thead>
                    <tbody
                      class="ant-table-tbody"
                    >
                      <tr
                        class="ant-table-row ant-table-row-level-0"
                        data-row-key="0"
                      >
                        <td
                          class="ant-table-cell"
                        >
                          <img
                            alt="track1 cover"
                            src=""
                          />
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          track1
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <button
                            class="ant-btn css-dev-only-do-not-override-12jzuas ant-btn-default"
                            type="button"
                          >
                            <span>
                              Enlever de Titres likés
                            </span>
                          </button>
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
